sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -220,9 +220,20 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
           attributeReference match {
             case attr @ FileSourceMetadataAttribute(
```

Comment:
I'm a bit confused. We remove all file source related metadata from the metadata col attribute in https://github.com/apache/spark/pull/40545/files#diff-4445cc3828e35092eb261467b499b8b0ef69ae694ea8ce25abf16b8ef4b72fbaR240 , how are we able to match it with `FileSourceMetadataAttribute` here?

Reply:
We only cleanup the individual metadata fields but the _metadata attribute itself does have the right information to match on `FileSourceMetadataAttribute` :
```
  def createFileMetadataCol(fileFormat: FileFormat): AttributeReference = {
    // Strip out the fields' metadata to avoid exposing it to the user. [[FileSourceStrategy]]
    // avoids confusion by mapping back to [[metadataSchemaFields]].
    val fields = metadataSchemaFields(fileFormat)
      .map(FileSourceMetadataAttribute.cleanupFileSourceMetadataInformation)
    FileSourceMetadataAttribute(FileFormat.METADATA_NAME, StructType(fields))
  }
```
The following is true:
```
val metadata_attr = createFileMetadataCol(fileFormat: FileFormat)
metadata_attr.metadata.getBoolean(METADATA_COL_ATTR_KEY) == true
metadata_attr.metadata.getBoolean(FILE_SOURCE_METADATA_COL_ATTR_KEY) == true

metadata_attr.dataType.forall(_.metadata == Metadata.empty)